### PR TITLE
IGNITE-18945 Add an ability to extend control-utility with the new commands

### PR DESCRIPTION
**Pull request by alex-plekhanov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/Command.java [hunk 1]
```diff
@@ -124,14 +124,34 @@ public default void usage(
...
+        String... args
     ) {
         logger.info("");
 
         if (experimental())
             logger.info(INDENT + "[EXPERIMENTAL]");
 
         logger.info(INDENT + desc);
-        logger.info(DOUBLE_INDENT + CommandLogger.join(" ", UTILITY_NAME, cmd, CommandLogger.join(" ", args)));
+        logger.info(DOUBLE_INDENT + CommandLogger.join(" ", UTILITY_NAME, cmd,
```
Comment (petrov-mg) [inline modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/Command.java:153]: Why this change?

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandArgIterator.java [hunk 2]
```diff
@@ -45,9 +49,14 @@ public class CommandArgIterator {
      * @param argsIt Raw argument iterator.
      * @param commonArgumentsAndHighLevelCommandSet All known subcomands.
      */
-    public CommandArgIterator(Iterator<String> argsIt, Set<String> commonArgumentsAndHighLevelCommandSet) {
+    public CommandArgIterator(
+        Iterator<String> argsIt,
+        Set<String> commonArgumentsAndHighLevelCommandSet,
+        Map<String, Command<?>> cmds
```
Comment (petrov-mg) [inline modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandArgIterator.java:55]: We need to describe the new parameter in javadoc.

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandHandler.java [hunk 3]
```diff
@@ -155,14 +160,34 @@ public static IgniteLogger setupJavaLogger(String appName, Class<?> cls) {
...
+
+        new CommandsProviderImpl().commands().forEach((k, v) -> cmds.put(k.toLowerCase(), v));
+
+        if (!F.isEmpty(it)) {
+            for (CommandsProvider provider : it) {
+                provider.commands().forEach((k, v) -> {
+                    k = k.toLowerCase();
+
+                    if (cmds.containsKey(k)) {
+                        throw new IllegalArgumentException("Only one action can be specified, but found at least two:" +
```
Comment (petrov-mg) [inline modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandHandler.java:183]: It seem that the error message should be changed - we need to say that command name conflict were found or something like that. And specifiy the plugin name/class in the error message if possible.

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandHandler.java [hunk 4]
```diff
@@ -155,14 +160,34 @@ public static IgniteLogger setupJavaLogger(String appName, Class<?> cls) {
...
     public CommandHandler(IgniteLogger logger) {
         this.logger = logger;
+        Iterable<CommandsProvider> it = U.loadService(CommandsProvider.class);
+
+        cmds = new HashMap<>();
+
+        new CommandsProviderImpl().commands().forEach((k, v) -> cmds.put(k.toLowerCase(), v));
+
+        if (!F.isEmpty(it)) {
+            for (CommandsProvider provider : it) {
```
Comment (petrov-mg) [inline modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandHandler.java:178]: It seems useful to add logging of what exeternal commands were registered.

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandsProviderImpl.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+package org.apache.ignite.internal.commandline;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Default implementation of CommandsProvider, based on {@link CommandList} enum.
+ */
+public class CommandsProviderImpl implements CommandsProvider {
```
Comment (petrov-mg) [inline modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandsProviderImpl.java:27]: It seems we can get rid of this class and move its logic to CommandList static method.

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/Command.java (hunk 1 above)
Comment (alex-plekhanov) [reply modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/Command.java]: Reverted

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandArgIterator.java (hunk 2 above)
Comment (alex-plekhanov) [reply modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandArgIterator.java:56]: Done

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandHandler.java (hunk 3 above)
Comment (alex-plekhanov) [reply modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandHandler.java]: Done

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandHandler.java (hunk 4 above)
Comment (alex-plekhanov) [reply modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandHandler.java:178]: There can be tools which parse the output of control.sh script and new unexpected output can break these tools. At least for "info" logging level. But we can use "debug" logging level.

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandsProviderImpl.java (hunk 5 above)
Comment (alex-plekhanov) [reply modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandsProviderImpl.java]: Done

**Review by petrov-mg — APPROVED** [review]
